internal/chain/engine.go

```diff
@@ -304,56 +307,42 @@ func (e *engineImpl) Finalize(
 		Uint64("block-number", header.Number().Uint64())
 
 	if isBeaconChain && inStakingEra {
-		superCommittee, err := chain.ReadShardState(
-			chain.CurrentHeader().Epoch(),
-		)
+		nowEpoch := chain.CurrentHeader().Epoch()
+		superCommittee, err := chain.ReadShardState(nowEpoch)
+		if err != nil {
+			return nil, nil, err
+		}
 		staked := superCommittee.StakedValidators()
 		// could happen that only harmony nodes are running,
-		if staked.CountStakedValidator > 0 {
-			l.RawJSON("external", []byte(staked.StateSubset.String())).
-				Msg("have non-zero external")
-
-			if err != nil {
-				return nil, nil, err
-			}
-
-			l.Msg("bumping validator signing counts")
-			if err := availability.IncrementValidatorSigningCounts(
-				chain, header, header.ShardID(), state, staked.LookupSet,
+		if isNewEpoch && staked.CountStakedValidator > 0 {
+			l.Msg("in new epoch (aka last block), apply availability check for activity")
+			if err := availability.SetInactiveUnavailableValidators(
```

Comment:
Seems it's too late to set inactive, those ones who should be kicked out are already proposed in shardState in the header. We should set inactive before shardState proposal